### ISSUE TITLE
Is it possible to use EF Core's DbSet<>.FromSqlRaw along with ToLinqToDB()?

### ISSUE THREAD

**Issue by djkunkel** [linq2db/linq2db.EntityFrameworkCore#37]
I have an application that bases some result sets on raw SQL queries for an entity. The IQueryable returned from DbSet<>.FromSqlRaw("select * from tablename") for example fails when attempting to enumerate the IQueryable then returned from ToLinqToDB(). Is this situation supported? I was hoping it would use a subquery to apply the additional filters added by users of the IQueryable returned from ToLinqToDB().

I'm using linq2db.EntityFrameworkCore 2.0.5, linq2db 2.9.7, EFCore 3.1.3 against SQL Server 2017.

I receive the following exception and stack trace:
```
The method or operation is not implemented.
  at Microsoft.EntityFrameworkCore.RelationalQueryableExtensions.FromSqlOnQueryable[TEntity](IQueryable`1 source, String sql, Object[] parameters)
```

  ```
 at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at LinqToDB.EntityFrameworkCore.LinqToDBForEFToolsImplDefault.<>c__DisplayClass35_0.<TransformExpression>g__LocalTransform|2(Expression e)
   at LinqToDB.EntityFrameworkCore.LinqToDBForEFToolsImplDefault.<>c__DisplayClass35_0.<TransformExpression>b__3(Expression e)
   at LinqToDB.Expressions.Extensions.Transform(Expression expr, Func`2 func)
   at LinqToDB.EntityFrameworkCore.LinqToDBForEFToolsImplDefault.TransformExpression(Expression expression, IDataContext dc, DbContext ctx, IModel model)
   at LinqToDB.EntityFrameworkCore.LinqToDBForEFTools.TransformExpression(Expression expression, IDataContext dc, DbContext ctx, IModel model)
   at LinqToDB.EntityFrameworkCore.LinqToDBForEFToolsDataContext.ProcessExpression(Expression expression)
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr)
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache)
   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at LinqToDB.EntityFrameworkCore.Internal.LinqToDBForEFQueryProvider`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Redacted.HealthCheckController.Get() in C:\redacted\HealthCheckController.cs:line 36
   at Microsoft.Extensions.Internal.ObjectMethodExecutor.Execute(Object target, Object[] parameters)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.SyncObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<<InvokeActionMethodAsync>g__Logged|12_1>d.MoveNext()
```

I'm calling a method on my data context that does the following:
```
public virtual IQueryable<CapacityData> GetRunsReport()
        {
            return CapacityRows.FromSqlRaw("select * from API_CapacityData_t where ProcessUnitCode IN (200, 201, 202, 203, 204, 205, 206)");
        }
```

and then, and an example, doing this with that context:
`_dc.GetRunsReport().ToLinqToDB().ToList();`

I understand this scenario might not be intended as supported, but in the case it is supposed to be handled, I'm opening this issue.

Thank you!

**Comment by sdanyliv** [comment]
Will find time to convert FromSqlRaw. We have our own `FromSql` method. Just forgot to create transformation.

**Comment by djkunkel** [comment]
I have a patch for this against version 2.0.5, do you accept pull requests? Also, I'm new to expression parsing and this project, so I may have done something terribly wrong, but I thought it might be helpful.

[0001-Support-FromSql-when-using-FromRawSql-from-EF.txt](https://github.com/linq2db/linq2db.EntityFrameworkCore/files/4484156/0001-Support-FromSql-when-using-FromRawSql-from-EF.txt)

**Comment by sdanyliv** [comment]
Which version of EF do you use?

**Comment by djkunkel** [comment]
3.1.3

**Comment by sdanyliv** [comment]
I suggest to update to linq2db.EntityFrameworkCore 3.0 - it has a lot of improvements, especially Eager Loading.

Anyway, you can customize current integration before I fix that without changing library code:
Implement your new class `MyImplementation` derived from `LinqToDBForEFToolsImplDefault` and override `TransformExpression` method.
And assign it to `LinqToDBForEFTools.Implementation = new MyImplementation()`

Sorry a little busy now, I'll update library in several days. Your changes looks good, but I need to verify and create tests.

**Comment by djkunkel** [comment]
Thanks for your feedback. I initially tried to override the implementation as suggested, but the changes were to an inline method (LocalTransform) inside the TransformExpression method.

**Comment by sdanyliv** [comment]
Hmm, strange, I thought that I have opened everything for cutomising. It should be easy, something like this:
```cs
public override Expression TransformExpression(Expression expression, IDataContext dc, DbContext ctx, IModel model)
{
   var expr = base.TransfromExpression(expression, dc, ctx, model);
   var expr = expr.Transform(e => {
      if ( e is MethodCallExpression and ... bla, bla is FromSqlRaw )
      {
         // your replace code
        return replaced;
      }
      return e;
   });

   return expr;
}
```
Anyway, will find a time to update library.

**Comment by djkunkel** [comment]
Thank you for this guidance. I was able to make my change in a custom override of TransformExpression!

**Comment by sdanyliv** [comment]
Released in 3.1.0
